File: ip-address-validation-nodejs/REST/get_geo_location_by_ipv4_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { IPAVResponse } from './ipav_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects IP Address Validation API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/GPP/web.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects IP Address Validation API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/GPP/web.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects IP Address Validation API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/GPP/web.svc/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.Number is not equal to '4'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.Number !== '4';

/**
 * <summary>
 * Constructs a full URL for the GetGeoLocationByIP_V4 API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}JSON/GetLocationByIP_V4?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<IPAVResponse>">A promise that resolves to an IPAVResponse object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
  try {
    const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
    return new IPAVResponse(response.data);
  } catch (error) {
    throw new Error(`HTTP request failed: ${error.message}`);
  }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects IP Address Validation API's GetGeoLocationByIP_V4 endpoint,
 * retrieving geographic location, proxy, host name, and US region information with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const GetGeoLocationByIPV4Client = {
    /**
     * <summary>
     * Asynchronously invokes the GetGeoLocationByIP_V4 API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.Number == '4') in live mode.
     * </summary>
     * @param {string} IPAddress - The IP address to look up, e.g., "209.85.173.104".
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<IPAVResponse>} - A promise that resolves to an IPAVResponse object.
     */
    async invokeAsync(IPAddress, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            IPAddress,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return isValid(fallbackResponse) ? fallbackResponse : response;
        }

        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the GetGeoLocationByIP_V4 API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @returns {IPAVResponse} - An IPAVResponse object with geographic location details or an error.
     */
    invoke(IPAddress, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            IPAddress, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { GetGeoLocationByIPV4Client, IPAVResponse };

File: ip-address-validation-python/REST/get_geo_location_by_ipv4_rest.py
from dataclasses import dataclass
from typing import Optional
import requests
from ipav_response import IPAVResponse, Error

# Endpoint URLs for IP Address Validation GetGeoLocationByIP_V4 REST API
primary_url = 'https://sws.serviceobjects.com/GPP/web.svc/JSON/GetLocationByIP_V4?'
backup_url = 'https://swsbackup.serviceobjects.com/GPP/web.svc/JSON/GetLocationByIP_V4?'
trial_url = 'https://trial.serviceobjects.com/GPP/web.svc/JSON/GetLocationByIP_V4?'

def get_geo_location_by_ipv4(
    ip_address: str,
    license_key: str,
    is_live: bool = True,
    timeout_seconds: int = 15
) -> IPAVResponse:
    """
    Call IP Address Validation GetGeoLocationByIP_V4 API to retrieve geographic location, proxy, host name, and US region information.

    Parameters:
        ip_address: The IP address to look up, e.g., "209.85.173.104".
        license_key: Your license key to use the service.
        is_live: Value to determine whether to use the live or trial servers (default: True).
        timeout_seconds: Timeout, in seconds, for the call to the service (default: 15).

    Returns:
        IPAVResponse: Parsed JSON response with geolocation information or error details.
    """
    params = {
        'IPAddress': ip_address,
        'LicenseKey': license_key
    }

    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url
    
    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=timeout_seconds)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error')
        if not (error is None or error.get('Number') != "4"):
          if is_live:
            # Try backup URL
            response = requests.get(backup_url, params=params, timeout=timeout_seconds)
            response.raise_for_status()
            data = response.json()

            # If still error, propagate exception
            if 'Error' in data:
                raise RuntimeError(f"IPAV service error: {data['Error']}")
          else:
              # Trial mode error is terminal
              raise RuntimeError(f"IPAV trial error: {data['Error']}")

        # Convert JSON response to IPAVResponse for structured access
        error = Error(**data.get('Error', {})) if data.get('Error') else None
        return IPAVResponse(
            IPAddress=data.get('IPAddress'),
            LicenseKey=data.get('LicenseKey'),
            Certainty=data.get('Certainty'),
            City=data.get('City'),
            Region=data.get('Region'),
            Country=data.get('Country'),
            CountryISO3=data.get('CountryISO3'),
            CountryISO2=data.get('CountryISO2'),
            PostalCode=data.get('PostalCode'),
            MetroCode=data.get('MetroCode'),
            DMA=data.get('DMA'),
            StateFIPS=data.get('StateFIPS'),
            CountyFIPS=data.get('CountyFIPS'),
            Latitude=data.get('Latitude'),
            Longitude=data.get('Longitude'),
            IsProxy=data.get('IsProxy'),
            ProxyType=data.get('ProxyType'),
            PossibleMobileDevice=data.get('PossibleMobileDevice'),
            ISP=data.get('ISP'),
            NetblockOwner=data.get('NetblockOwner'),
            HostNames=data.get('HostNames'),
            IPNoteCodes=data.get('IPNoteCodes'),
            IPNotes=data.get('IPNotes'),
            Debug=data.get('Debug'),
            Error=error
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                data = response.json()
                if 'Error' in data:
                    raise RuntimeError(f"IPAV backup error: {data['Error']}") from req_exc

                # Convert JSON response to IPAVResponse for structured access
                error = Error(**data.get('Error', {})) if data.get('Error') else None
                return IPAVResponse(
                    IPAddress=data.get('IPAddress'),
                    LicenseKey=data.get('LicenseKey'),
                    Certainty=data.get('Certainty'),
                    City=data.get('City'),
                    Region=data.get('Region'),
                    Country=data.get('Country'),
                    CountryISO3=data.get('CountryISO3'),
                    CountryISO2=data.get('CountryISO2'),
                    PostalCode=data.get('PostalCode'),
                    MetroCode=data.get('MetroCode'),
                    DMA=data.get('DMA'),
                    StateFIPS=data.get('StateFIPS'),
                    CountyFIPS=data.get('CountyFIPS'),
                    Latitude=data.get('Latitude'),
                    Longitude=data.get('Longitude'),
                    IsProxy=data.get('IsProxy'),
                    ProxyType=data.get('ProxyType'),
                    PossibleMobileDevice=data.get('PossibleMobileDevice'),
                    ISP=data.get('ISP'),
                    NetblockOwner=data.get('NetblockOwner'),
                    HostNames=data.get('HostNames'),
                    IPNoteCodes=data.get('IPNoteCodes'),
                    IPNotes=data.get('IPNotes'),
                    Debug=data.get('Debug'),
                    Error=error
                )
            except Exception as backup_exc:
                raise RuntimeError("IPAV service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"IPAV trial error: {str(req_exc)}") from req_exc
